src/app/clusters/energy-evse-server/energy-evse-server.cpp

```diff
@@ -360,27 +362,84 @@ void Instance::HandleStartDiagnostics(HandlerContext & ctx, const Commands::Star
 {
-    // Call the delegate
-    // TODO
-    // Status status = mDelegate.GetTargets();
-    Status status = Status::UnsupportedCommand;
 
-    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+    CHIP_ERROR err                               = CHIP_NO_ERROR;
+    app::CommandHandler::Handle commandHandleRef = CommandHandler::Handle(&ctx.mCommandHandler);
+    auto commandHandle                           = commandHandleRef.Get();
+
+    TLV::TLVWriter * writer;
+    TLV::TLVType listContainerType;
+    EvseTargetIterator * it = nullptr;
+
+    EvseTargetEntry chargingTargetScheduleEntry;
+    Structs::ChargingTargetScheduleStruct::Type entry;
+    Structs::ChargingTargetStruct::Type array[kEvseTargetsMaxTargetsPerDay];
+
+    if (commandHandle == nullptr)
+    {
+        // When the platform has shut down, interaction model engine will invalidate all commandHandle to avoid dangling references.
+        return;
+    }
+
+    const CommandHandler::InvokeResponseParameters prepareParams(ctx.mRequestPath);
+    SuccessOrExit(
+        err = commandHandle->PrepareInvokeResponseCommand(
+            ConcreteCommandPath(ctx.mRequestPath.mEndpointId, ctx.mRequestPath.mClusterId, Commands::GetTargetsResponse::Id),
+            prepareParams));
+
+    VerifyOrExit((writer = commandHandle->GetCommandDataIBTLVWriter()) != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    SuccessOrExit(err = writer->StartContainer(TLV::ContextTag(Commands::GetTargetsResponse::Fields::kChargingTargetSchedules),
+                                               TLV::TLVType::kTLVType_Array, listContainerType));
+
+    SuccessOrExit(err = mDelegate.PrepareGetTargets(&it));
+    VerifyOrExit(it != nullptr, err = CHIP_ERROR_UNINITIALIZED);
+
+    while (it->Next(chargingTargetScheduleEntry))
+    {
+        entry.dayOfWeekForSequence = chargingTargetScheduleEntry.dayOfWeekMap;
+
+        uint8_t index = 0;
+        // copy from our vector to a local array of structs which we can reduce to the length required
+        // so we can pass it into our Encode engine
+        for (const EvseChargingTarget & chargingTarget : chargingTargetScheduleEntry.dailyChargingTargets)
+        {
+            array[index].targetTimeMinutesPastMidnight = chargingTarget.targetTimeMinutesPastMidnight;
+            array[index].targetSoC                     = chargingTarget.targetSoC;
+            array[index].addedEnergy                   = chargingTarget.addedEnergy;
+            index++;
+        }
+
+        DataModel::List<Structs::ChargingTargetStruct::Type> mSpan(array);
+        mSpan.reduce_size(index);
+        entry.chargingTargets = mSpan;
+
+        SuccessOrExit(err = DataModel::Encode(*writer, TLV::AnonymousTag(), entry));
+    }
+
+    SuccessOrExit(err = writer->EndContainer(listContainerType));
+    SuccessOrExit(err = commandHandle->FinishCommand());
+exit:
+    // Release iterator and any memory
```

Comment:
So if something above errors out after PrepareInvokeResponseCommand succeeds but before FinishCommand, doesn't this leave the CommandHandler in a bad state and not properly respond with an error?  This code also does not work with command batching.

The raw APIs are there, but using them correctly is ... not trivial.

I strongly suggest using `TryAddingResponse` to at least not leave things in a broken state, though the "respond with error" part will still be on you.

The other option is to define a class with an `Encode()` method that does the right things to encode what you want encoded and then you can use the `AddResponse()` API on CommandHandler which does all the sanity bits for you.

Reply:
I have changed the code so in the error condition the following happens:

`ctx.mCommandHandler.AddStatus(ctx.mRequestPath, StatusIB(err).mStatus);`

Is that ok?

Reply:
I have updated from master and had to rework the code as a consequence of https://github.com/project-chip/connectedhomeip/pull/33736 I believe. 

Would it be possible to review the reworked code pls? 

In particular is the 
`    EnergyEvse::Structs::ChargingTargetStruct::Type chargingTargets[kEvseTargetsMaxNumberOfDays][kEvseTargetsMaxTargetsPerDay];
`
acceptable? I have run tests and they passed. I have assumed that 
`    ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
`copies the chargingTargets pointed to in the Lists so that when the method returns and `chargingTargets `goes out of scope all is ok?